service/src/main/java/bio/terra/workspace/service/resource/model/WsmResourceState.java

```diff
@@ -0,0 +1,110 @@
+package bio.terra.workspace.service.resource.model;
+
+import bio.terra.common.exception.ValidationException;
+import bio.terra.workspace.common.exception.InternalLogicException;
+import bio.terra.workspace.generated.model.ApiState;
+import java.util.List;
+import java.util.Map;
+import org.apache.commons.lang3.SerializationException;
+
+/**
+ * See design document <a
+ * href="https://docs.google.com/document/d/1ply9gRK-X8luKOrCYbZAA90QGqoZwjyukCsrA5PbIyQ/edit#heading=h.ygdsvwmy6va0">here</a>
+ */
+public enum WsmResourceState {
+  BROKEN("BROKEN", ApiState.BROKEN),
+  BROKEN_DELETE("BROKEN_DELETE", ApiState.BROKEN_DELETE),
+  CREATING("CREATING", ApiState.CREATING),
+  DELETING("DELETING", ApiState.DELETING),
+  READY("READY", ApiState.READY),
+  UPDATING("UPDATING", ApiState.UPDATING),
+  NOT_EXISTS("invalid not exists", null);
+
+  // -- Valid state transitions --
+  private static final List<WsmResourceState> BROKEN_TRANSITIONS = List.of(DELETING);
+  // BROKEN_DELETING requires manual intervention to clean up
+  private static final List<WsmResourceState> BROKEN_DELETING_TRANSITIONS = List.of();
+  // We allow the NOT_EXISTS transition for the V1 apis where we move failed resources from
+  // creating to non-existent.
+  private static final List<WsmResourceState> CREATING_TRANSITIONS =
+      List.of(BROKEN, READY, NOT_EXISTS);
```

Comment:
to check my understanding, for v1, they will all be NOT_EXIST when creation fail but for v2, we will keep the resource in the db with a broken state?

Reply:
Right, exactly. I want to move away from the "v1" and "v2", so I updated the comment to refer to the state rule.